### PR TITLE
MDL-72321 question: Implement question filtering v3

### PR DESCRIPTION
**Pull request by TomoTsuyuki**
#311 #312 

This is just rebased and cleaned commits from old version => #322 

*** PLEASE DO NOT OPEN PULL REQUESTS VIA GITHUB ***

The moodle.git repository at Github is just a mirror of the official repository. We do not accept pull requests at Github.

See CONTRIBUTING.txt guidelines for how to contribute patches for Moodle. Thank you.

--

### CODE REVIEW

File: question/classes/external/bank.php [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        $questionbank->set_pagevars($params);
+        $questionbank->add_standard_searchcondition();
+        ob_start();
+        $questionbank->display_for_api();
+        $tablehtml = ob_get_clean();
+
+        $totalquestions = $questionbank->get_question_count();
+
+        return [
+            'html' => $tablehtml,
```
[the remark this answers is not in the document]
Comment (TomoTsuyuki) [reply question/classes/external/bank.php]: Updated here to reply only id, contextid, and name, and use fragment to show html now.

**Comment by TomoTsuyuki** [comment]
The branch has been rebased by latest filter codes by #349 
I updated to use fragment now, but haven't changed filter object and security check.
I updated phpunit test, but not checked behat test yet.

**Comment by mattporritt** [comment]
This looks ok, let me know when you've done the remaing tasks and I'll give it another look

File: question/classes/external/bank.php [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+                VALUE_DEFAULT,
+                0
+            ),
+            'qbshowtext' => new external_value(
+                PARAM_BOOL,
+                'Flag to show question text',
+                VALUE_DEFAULT,
+                false,
+            ),
+            'recurse' => new external_value(
```
[the remark this answers is not in the document]
Comment (TomoTsuyuki) [reply question/classes/external/bank.php]: I moved `filterverb`, `showhidden`, and `recurse` into `filteroptions`.
Also moved `perpage`, `showtext`, and `page` into `displayoptions`.

File: question/classes/external/bank.php [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        $categoryids = $params['filters']['category']['values'];
+        // Currently, we support only one category for the list because of new/edit/delete buttons.
+        $categoryid = array_pop($categoryids);
+        $categories = $DB->get_records('question_categories', ['id' => $categoryid]);
+        $categories = \qbank_managecategories\helper::question_add_context_in_key($categories);
+        $category = array_pop($categories);
+        $category = $category->id;
+        $params['cat'] = $category;
+
+        require_login($courseid, false);
```
[the remark this answers is not in the document]
Comment (TomoTsuyuki) [reply question/classes/external/bank.php:181]: I found the capability check for the view https://github.com/moodle/moodle/blob/d135a1200a87a62cad7afa2e51889b845f72e4cb/question/editlib.php#L311 so I use same thing here.
Currently we can search by one category so I believe that's enough for now.

**Comment by TomoTsuyuki** [comment]
I updated to use filteroptions/displayoptions, and add security check.
I'm fixing the behat test below:
vendor/bin/behat --config /var/www/behatdata/behatrun/behat/behat.yml --tags="@core_question"

**Comment by safatshahin** [comment]
closing this as this issue is now work in progress in a different branch.

**Pull request closed without merging**